panda_agent/server.py

```diff
@@ -40,50 +57,172 @@ def StartAgent(self, request: pb.StartAgentRequest, context):
+
+        Returns:
+            gRPC response containing the recording name and the filename of the recording files
+
+        Raises:
+            RuntimeError if PANDA was not running
+            RuntimeWarning if PANDA was not recording
+        '''
         if self.agent is not None:
-            raise RuntimeError(ErrorCode.RUNNING.value, "Cannot start another instance of PANDA while one is already running")
+            recording_name = self.agent.stop_recording()
+            return pb.StopRecordingResponse(
+                recording_name=recording_name,
+                snapshot_filename=f"{recording_name}-rr-snp",
+                ndlog_filename=f"{recording_name}-rr-nondet.log"
+            )
+        else:
+            raise RuntimeError(ErrorCode.NOT_RUNNING, "Cannot stop a recording when PANDA is not running")
+
+    def StartReplay(self, request: pb.StartReplayRequest, context):
+        '''
+        Starts a PANDA replay with the specified configuration.
+
+        Only one PANDA instance per agent can be running at one time.
+
+        Args:
+            request: gRPC request containing the replay name
+
+        Returns:
+            gRPC response containing the serial and PANDA output of the replay
+
+        Raises:
+            RuntimeError if PANDA was already running
+            RuntimeError if recording files to replay to not exist
+        '''
+        if self.agent is not None and self.agent.panda.started.is_set():
+            raise RuntimeError(ErrorCode.RUNNING, "Cannot start another instance of PANDA while one is already running")
         
+        print("Starting replay")
+        self.agent = PandaAgent(request.config)
+
         serial = self.agent.start_replay(request.recording_name)
-        with (open("./shared/execution.log")) as file:
-            replay = file.read()
-        return pb.StartReplayResponse(serial=serial, replay=replay)
+        log = read_execution_log()
+        
+        return pb.StartReplayResponse(serial=serial, replay=log)
 
     def StopReplay(self, request: pb.StopReplayRequest, context):
-        serial = self.agent.stop_replay()
-        with (open("./shared/execution.log")) as file:
-            replay = file.read()
-        return pb.StopReplayResponse(serial=serial, replay=replay)
+        '''
+        Stops a PANDA replay.
+
+        PANDA replays stop naturally.
+
+        Args:
+            request: gRPC request
+
+        Returns:
+            gRPC response containing the serial and PANDA output of the replay
+
+        Raises:
+            RuntimeError if PANDA was not replaying
+        '''
+        if self.agent is not None:    
+            serial = self.agent.stop_replay()
+            log = read_execution_log()
+            return pb.StopReplayResponse(serial=serial, replay=log)
+        else:
+            raise RuntimeError(ErrorCode.NOT_REPLAYING, "Must start a replay before stopping one")
 
     def SendNetworkCommand(self, request: pb.NetworkRequest, context):
```

Comment:
I will write a bit of a better description for this one specifically, but overall the current documentation is fine. I just want to be pedantic and slightly change this but don't now to what yet.